### PR TITLE
exclude rooms having undefined reportID

### PR DESCRIPTION
**Pull request by sourcecodedeveloper**
### Fixed Issues

$ [#30381](https://github.com/Expensify/App/issues/30381)
PROPOSAL: https://github.com/Expensify/App/issues/30381#issuecomment-1781953824

### Tests

The test scenario applies to Web only.  The PR would resolve the error on other platforms as well.
1.  Login using Account A and create workspace room
2.  Open another tab and login using Account A.
3.  Sign out of the first tab, takes few tries, but eventually you should get blank screen and signed out yet still able to click on the plus icon. Click on the plus icon to create new workspace.
4.  Now sign back into the app using Account A.
5.  Navigate to workspace and notice the error "your session has expired please sign in again" with grayed out workspace.
6. Click on the grayed out workspace and then click on threedots menu. 
7. Notice the only available option is 'Delete Workspace'. The test confirms that the rooms #announce and #admins are not visible because the reportID is undefined.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

[web.webm](https://github.com/Expensify/App/assets/141340971/b4827e3b-f71f-4a9f-8837-47f7389068b5)

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by roryabraham** [comment]
Tagging @allroundexperts for the C+ review

**Comment by sourcecodedeveloper** [comment]
> @sourcecodedeveloper Your checklist is not complete. Please check off all the items.

@allroundexperts PR is ready for review.

**Comment by sourcecodedeveloper** [comment]
> @sourcecodedeveloper Sorry, what is this new commit about?

Please ignore, reverting.

**Comment by sourcecodedeveloper** [comment]
@allroundexperts Reassure Performance Tests are passing on local branch. Probably just need to re-run workflow.
Test Suites: 7 passed, 7 total
Tests:       18 passed, 18 total
Snapshots:   0 total

**Comment by sourcecodedeveloper** [comment]
@roryabraham awaiting to finish review and merge.

**Comment by roryabraham** [comment]
ReportScreen reassure tests are known to be flaky. We'll try one more time and if it's still failing we'll merge anyways

**Review by roryabraham — CHANGES_REQUESTED** [review]
This component has poor memoization. For context, it's not unheard of for some users to have dozens or hundreds of policies and tens of thousands of reports.

This code will loop through all reports a user has access to on every render, plus any time any field of the policy changes, plus any time you click on one of the menu items. Let's implement some smarter memoization in this component:

```diff
diff --git a/src/pages/workspace/WorkspaceInitialPage.js b/src/pages/workspace/WorkspaceInitialPage.js
index 5153a30a28..1a94288e2f 100644
--- a/src/pages/workspace/WorkspaceInitialPage.js
+++ b/src/pages/workspace/WorkspaceInitialPage.js
@@ -1,6 +1,6 @@
 import lodashGet from 'lodash/get';
 import PropTypes from 'prop-types';
-import React, {useCallback, useEffect, useState} from 'react';
+import React, {useCallback, useEffect, useMemo, useState} from 'react';
 import {ScrollView, View} from 'react-native';
 import {withOnyx} from 'react-native-onyx';
 import _ from 'underscore';
@@ -15,10 +15,10 @@ import PressableWithoutFeedback from '@components/Pressable/PressableWithoutFeed
 import ScreenWrapper from '@components/ScreenWrapper';
 import Text from '@components/Text';
 import Tooltip from '@components/Tooltip';
-import withLocalize, {withLocalizePropTypes} from '@components/withLocalize';
-import withWindowDimensions from '@components/withWindowDimensions';
+import useLocalize from '@hooks/useLocalize';
 import useSingleExecution from '@hooks/useSingleExecution';
 import useWaitForNavigation from '@hooks/useWaitForNavigation';
+import useWindowDimensions from '@hooks/useWindowDimensions';
 import compose from '@libs/compose';
 import Navigation from '@libs/Navigation/Navigation';
 import * as PolicyUtils from '@libs/PolicyUtils';
@@ -37,7 +37,6 @@ import withPolicyAndFullscreenLoading from './withPolicyAndFullscreenLoading';
 
 const propTypes = {
     ...policyPropTypes,
-    ...withLocalizePropTypes,
 
     /** All reports shared with the user (coming from Onyx) */
     reports: PropTypes.objectOf(reportPropTypes),
@@ -74,17 +73,45 @@ function WorkspaceInitialPage(props) {
     const hasPolicyCreationError = Boolean(policy.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD && policy.errors);
     const waitForNavigate = useWaitForNavigation();
     const {singleExecution, isExecuting} = useSingleExecution();
+    const {translate} = useLocalize();
+    const {windowWidth} = useWindowDimensions();
+
+    const policyID = useMemo(() => policy.id, [policy]);
+    const [policyReports, adminsRoom, announceRoom] = useMemo(() => {
+        const reports = [];
+        let admins;
+        let announce;
+        _.each(props.reports, (report) => {
+            if (!report || report.policyID !== policyID) {
+                return;
+            }
+            reports.push(report);
+
+            if (!ReportUtils.isThread(report)) {
+                return;
+            }
+
+            if (report.chatType === CONST.REPORT.CHAT_TYPE.POLICY_ADMINS) {
+                admins = report;
+                return;
+            }
+
+            if (report.chatType === CONST.REPORT.CHAT_TYPE.POLICY_ANNOUNCE) {
+                announce = report;
+            }
+        });
+        return [reports, admins, announce];
+    }, [policyID, props.reports]);
 
     /**
      * Call the delete policy and hide the modal
      */
     const confirmDeleteAndHideModal = useCallback(() => {
-        const policyReports = _.filter(props.reports, (report) => report && report.policyID === policy.id);
         Policy.deleteWorkspace(policy.id, policyReports, policy.name);
         setIsDeleteModalOpen(false);
         // Pop the deleted workspace page before opening workspace settings.
         Navigation.goBack(ROUTES.SETTINGS_WORKSPACES);
-    }, [props.reports, policy]);
+    }, [policy.id, policy.name, policyReports]);
 
     useEffect(() => {
         const policyDraftId = lodashGet(props.policyDraft, 'id', null);
@@ -111,19 +138,7 @@ function WorkspaceInitialPage(props) {
         Policy.updateGeneralSettings(policy.id, policy.name, CONST.CURRENCY.USD);
         setIsCurrencyModalOpen(false);
         ReimbursementAccount.navigateToBankAccountRoute(policy.id);
-    }, [policy]);
-
-    /**
-     * Navigates to workspace rooms
-     * @param {String} chatType
-     */
-    const goToRoom = useCallback(
-        (type) => {
-            const room = _.find(props.reports, (report) => report && report.policyID === policy.id && report.chatType === type && !ReportUtils.isThread(report));
-            Navigation.dismissModal(room.reportID);
-        },
-        [props.reports, policy],
-    );
+    }, [policy.id, policy.name]);
 
     const policyName = lodashGet(policy, 'name', '');
     const hasMembersError = PolicyUtils.hasPolicyMemberError(props.policyMembers);
@@ -179,32 +194,30 @@ function WorkspaceInitialPage(props) {
         },
     ];
 
-    const availableRooms = _.filter(
-        props.reports,
-        (report) =>
-            report &&
-            report.reportID !== undefined &&
-            report.policyID === policy.id &&
-            (report.chatType === CONST.REPORT.CHAT_TYPE.POLICY_ADMINS || report.chatType === CONST.REPORT.CHAT_TYPE.POLICY_ANNOUNCE),
-    );
-
-    const threeDotsMenuItems = [
-        {
-            icon: Expensicons.Trashcan,
-            text: props.translate('workspace.common.delete'),
-            onSelected: () => setIsDeleteModalOpen(true),
-        },
-    ];
-
-    availableRooms.forEach((room) => {
-        threeDotsMenuItems.push({
-            icon: Expensicons.Hashtag,
-            text: props.translate('workspace.common.goToRoom', {
-                roomName: room.chatType === CONST.REPORT.CHAT_TYPE.POLICY_ADMINS ? CONST.REPORT.WORKSPACE_CHAT_ROOMS.ADMINS : CONST.REPORT.WORKSPACE_CHAT_ROOMS.ANNOUNCE,
-            }),
-            onSelected: () => goToRoom(room.chatType),
-        });
-    });
+    const threeDotsMenuItems = useMemo(() => {
+        const items = [
+            {
+                icon: Expensicons.Trashcan,
+                text: translate('workspace.common.delete'),
+                onSelected: () => setIsDeleteModalOpen(true),
+            },
+        ];
+        if (adminsRoom) {
+            items.push({
+                icon: Expensicons.Hashtag,
+                text: translate('workspace.common.goToRoom', {roomName: CONST.REPORT.WORKSPACE_CHAT_ROOMS.ADMINS}),
+                onSelected: () => Navigation.dismissModal(adminsRoom.reportID),
+            });
+        }
+        if (announceRoom) {
+            items.push({
+                icon: Expensicons.Hashtag,
+                text: translate('workspace.common.goToRoom', {roomName: CONST.REPORT.WORKSPACE_CHAT_ROOMS.ANNOUNCE}),
+                onSelected: () => Navigation.dismissModal(adminsRoom.reportID),
+            });
+        }
+        return items;
+    }, [adminsRoom, announceRoom, translate]);
 
     return (
         <ScreenWrapper
@@ -218,7 +231,7 @@ function WorkspaceInitialPage(props) {
                     subtitleKey={_.isEmpty(policy) ? undefined : 'workspace.common.notAuthorized'}
                 >
                     <HeaderWithBackButton
-                        title={props.translate('workspace.common.workspace')}
+                        title={translate('workspace.common.workspace')}
                         shouldShowThreeDotsButton
                         shouldShowGetAssistanceButton
                         singleExecution={singleExecution}
@@ -226,7 +239,7 @@ function WorkspaceInitialPage(props) {
                         shouldDisableThreeDotsButton={isExecuting}
                         guidesCallTaskID={CONST.GUIDES_CALL_TASK_IDS.WORKSPACE_INITIAL}
                         threeDotsMenuItems={threeDotsMenuItems}
-                        threeDotsAnchorPosition={styles.threeDotsPopoverOffset(props.windowWidth)}
+                        threeDotsAnchorPosition={styles.threeDotsPopoverOffset(windowWidth)}
                         onBackButtonPress={() => Navigation.goBack(ROUTES.SETTINGS_WORKSPACES)}
                     />
                     <ScrollView contentContainerStyle={[styles.flexGrow1, styles.flexColumn, styles.justifyContentBetween, safeAreaPaddingBottomStyle]}>
@@ -239,12 +252,12 @@ function WorkspaceInitialPage(props) {
                             <View style={[styles.flex1]}>
                                 <View style={styles.avatarSectionWrapper}>
                                     <View style={[styles.settingsPageBody, styles.alignItemsCenter]}>
-                                        <Tooltip text={props.translate('workspace.common.settings')}>
+                                        <Tooltip text={translate('workspace.common.settings')}>
                                             <PressableWithoutFeedback
                                                 disabled={hasPolicyCreationError || isExecuting}
                                                 style={[styles.pRelative, styles.avatarLarge]}
                                                 onPress={singleExecution(waitForNavigate(() => openEditor(policy.id)))}
-                                                accessibilityLabel={props.translate('workspace.common.settings')}
+                                                accessibilityLabel={translate('workspace.common.settings')}
                                                 accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
                                             >
                                                 <Avatar
@@ -259,12 +272,12 @@ function WorkspaceInitialPage(props) {
                                             </PressableWithoutFeedback>
                                         </Tooltip>
                                         {!_.isEmpty(policy.name) && (
-                                            <Tooltip text={props.translate('workspace.common.settings')}>
+                                            <Tooltip text={translate('workspace.common.settings')}>
                                                 <PressableWithoutFeedback
                                                     disabled={hasPolicyCreationError || isExecuting}
                                                     style={[styles.alignSelfCenter, styles.mt4, styles.w100]}
                                                     onPress={singleExecution(waitForNavigate(() => openEditor(policy.id)))}
-                                                    accessibilityLabel={props.translate('workspace.common.settings')}
+                                                    accessibilityLabel={translate('workspace.common.settings')}
                                                     accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
                                                 >
                                                     <Text
@@ -287,7 +300,7 @@ function WorkspaceInitialPage(props) {
                                         key={item.translationKey}
                                         disabled={hasPolicyCreationError || isExecuting}
                                         interactive={!hasPolicyCreationError}
-                                        title={props.translate(item.translationKey)}
+                                        title={translate(item.translationKey)}
                                         icon={item.icon}
                                         iconRight={item.iconRight}
                                         onPress={item.action}
@@ -299,23 +312,23 @@ function WorkspaceInitialPage(props) {
                         </OfflineWithFeedback>
                     </ScrollView>
                     <ConfirmModal
-                        title={props.translate('workspace.bankAccount.workspaceCurrency')}
+                        title={translate('workspace.bankAccount.workspaceCurrency')}
                         isVisible={isCurrencyModalOpen}
                         onConfirm={confirmCurrencyChangeAndHideModal}
                         onCancel={() => setIsCurrencyModalOpen(false)}
-                        prompt={props.translate('workspace.bankAccount.updateCurrencyPrompt')}
-                        confirmText={props.translate('workspace.bankAccount.updateToUSD')}
-                        cancelText={props.translate('common.cancel')}
+                        prompt={translate('workspace.bankAccount.updateCurrencyPrompt')}
+                        confirmText={translate('workspace.bankAccount.updateToUSD')}
+                        cancelText={translate('common.cancel')}
                         danger
                     />
                     <ConfirmModal
-                        title={props.translate('workspace.common.delete')}
+                        title={translate('workspace.common.delete')}
                         isVisible={isDeleteModalOpen}
                         onConfirm={confirmDeleteAndHideModal}
                         onCancel={() => setIsDeleteModalOpen(false)}
-                        prompt={props.translate('workspace.common.deleteConfirmation')}
-                        confirmText={props.translate('common.delete')}
-                        cancelText={props.translate('common.cancel')}
+                        prompt={translate('workspace.common.deleteConfirmation')}
+                        confirmText={translate('common.delete')}
+                        cancelText={translate('common.cancel')}
                         danger
                     />
                 </FullPageNotFoundView>
@@ -329,9 +342,7 @@ WorkspaceInitialPage.defaultProps = defaultProps;
 WorkspaceInitialPage.displayName = 'WorkspaceInitialPage';
 
 export default compose(
-    withLocalize,
     withPolicyAndFullscreenLoading,
-    withWindowDimensions,
     withOnyx({
         reports: {
             key: ONYXKEYS.COLLECTION.REPORT,

```

**Comment by sourcecodedeveloper** [comment]
@roryabraham Added memoization, also that goToRoom callback was not necessary.

The PR Author Checklist has an error 'Resource not accessible by integration'

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 1]
```diff
@@ -75,16 +74,42 @@ function WorkspaceInitialPage(props) {
...
      * Call the delete policy and hide the modal
      */
     const confirmDeleteAndHideModal = useCallback(() => {
-        const policyReports = _.filter(props.reports, (report) => report && report.policyID === policy.id);
         Policy.deleteWorkspace(policy.id, policyReports, policy.name);
         setIsDeleteModalOpen(false);
         // Pop the deleted workspace page before opening workspace settings.
         Navigation.goBack(ROUTES.SETTINGS_WORKSPACES);
-    }, [props.reports, policy]);
+    }, [policy.id, policy.name, policyReports]);
```
Comment (roryabraham) [inline src/pages/workspace/WorkspaceInitialPage.js:112]:
```suggestion
    }, [policyID, policy.name, policyReports]);
```

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 2]
```diff
@@ -111,19 +136,7 @@ function WorkspaceInitialPage(props) {
...
-     * @param {String} chatType
-     */
-    const goToRoom = useCallback(
-        (type) => {
-            const room = _.find(props.reports, (report) => report && report.policyID === policy.id && report.chatType === type && !ReportUtils.isThread(report));
-            Navigation.dismissModal(room.reportID);
-        },
-        [props.reports, policy],
-    );
+    }, [policy.id, policy.name]);
```
Comment (roryabraham) [inline src/pages/workspace/WorkspaceInitialPage.js:139]:
```suggestion
    }, [policyID, policy.name]);
```

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 3]
```diff
@@ -111,19 +136,7 @@ function WorkspaceInitialPage(props) {
         Policy.updateGeneralSettings(policy.id, policy.name, CONST.CURRENCY.USD);
```
Comment (roryabraham) [inline src/pages/workspace/WorkspaceInitialPage.js:136]:
```suggestion
        Policy.updateGeneralSettings(policyID, policy.name, CONST.CURRENCY.USD);
```

**Review by roryabraham — CHANGES_REQUESTED** [review]
Thanks so much for making those changes. This is looking a lot better. Just a few more small things I noticed

**Review by roryabraham — APPROVED** [review]
LGTM, but we need to retest before merging as there have been substantial changes

**Comment by roryabraham** [comment]
PR Author Checklist failure is unrelated – will be fixed by https://github.com/Expensify/App/pull/31233

**Comment by sourcecodedeveloper** [comment]
> LGTM, but we need to retest before merging as there have been substantial changes

Agreed, I did retesting before I pushed changes.

**Comment by roryabraham** [comment]
Everything appears to be working as expected:

image

Thanks @sourcecodedeveloper for making those changes

**Pull request merged**